src/ucai/core/utils/callable_utils.py

```diff
@@ -0,0 +1,285 @@
+            )
+        elif origin == dict:
+            return (
+                f"Error in parameter '{param_name}': Dict type requires both key and value types. "
+                f"Please define the internal types for the dict, e.g., Dict[str, int]. Original error: {exception}"
+            )
+    return (
+        f"Error in parameter '{param_name}': type {param_hint} is not supported. "
+        f"Original error: {exception}"
+    )
+
+def format_default_value(default: Any) -> str:
+    """Formats the default value for SQL."""
+    if default is None:
+        return 'NULL'
+    elif isinstance(default, str):
+        return f"'{default}'"
+    else:
+        return str(default)
+
+def unwrap_function(func: Callable[..., Any]) -> Callable[..., Any]:
+    """Unwraps staticmethod or classmethod to get the actual function."""
+    if isinstance(func, (staticmethod, classmethod)):
+        func = func.__func__
+    return func
+
+def process_parameter(
+    param_name: str,
+    param: inspect.Parameter,
+    type_hints: dict[str, Any],
+    docstring_comments: dict[str, str]
+) -> str:
+    """Processes a single parameter and returns its SQL definition."""
+    if param_name in FORBIDDEN_PARAMS:
+        raise ValueError(f"Parameter '{param_name}' is not allowed in the function signature.")
+
+    if param.kind in (inspect.Parameter.VAR_POSITIONAL, inspect.Parameter.VAR_KEYWORD):
+        kind = 'var-positional (*args)' if param.kind == inspect.Parameter.VAR_POSITIONAL else 'var-keyword (**kwargs)'
+        raise ValueError(f"Parameter '{param_name}' is a {kind} parameter, which is not supported in SQL functions.")
+
+    if param_name not in type_hints:
+        raise ValueError(f"Missing type hint for parameter: {param_name}.")
+
+    param_hint = type_hints[param_name]
+
+    try:
+        sql_type = validate_type_hint(param_hint)
+    except ValueError as e:
+        error_message = generate_type_hint_error_message(param_name, param_hint, e)
+        raise ValueError(error_message) from e
+
+    param_comment = docstring_comments.get(param_name, f"Parameter {param_name}").replace("'", '"')
+
+    if param.default is inspect.Parameter.empty:
+        return f"{param_name} {sql_type} COMMENT '{param_comment}'"
+    else:
+        default_value = format_default_value(param.default)
+        return f"{param_name} {sql_type} DEFAULT {default_value} COMMENT '{param_comment}'"
+
+def assemble_sql_body(
+    catalog: str,
+    schema: str,
+    func_name: str,
+    sql_params: list[str],
+    sql_return_type: str,
+    func_comment: str,
+    indented_body: str
+) -> str:
+    """Assembles the final SQL function body."""
+    sql_params_str = ', '.join(sql_params)
+    sql_body = f"""
+    CREATE OR REPLACE FUNCTION {catalog}.{schema}.{func_name}({sql_params_str})
+    RETURNS {sql_return_type}
+    LANGUAGE PYTHON
+    COMMENT '{func_comment}'
+    AS $$
+{indented_body}
+    $$;
+    """
```

Comment:
```suggestion
    sql_body = f"""
CREATE OR REPLACE FUNCTION {catalog}.{schema}.{func_name}({sql_params_str})
RETURNS {sql_return_type}
LANGUAGE PYTHON
COMMENT '{func_comment}'
AS $$
{indented_body}
$$;
"""
```

should'n we dedent?

Reply:
yep - updated all of the tests and I'm preserving the indentation of the body to align with the expected layout for UC python functions. Also confirmed that it doesn't matter about dedent / indent on the python function that is registered, provided that relative indentation is consistent and correct for the function body. The current implementation matches the tutorials here: https://docs.databricks.com/en/sql/language-manual/sql-ref-syntax-ddl-create-sql-function.html#create-python-functions